statistics/histogram.go

```diff
@@ -686,6 +690,71 @@ func (hg *Histogram) outOfRange(val types.Datum) bool {
 		chunk.Compare(hg.Bounds.GetRow(hg.Bounds.NumRows()-1), 0, &val) < 0
 }
 
+// MergeHistogramForInnerJoin merges two histogram into one for inner join.
+func MergeHistogramForInnerJoin(lSide *Histogram, rSide *Histogram, tp *types.FieldType) *Histogram {
+	cmpFunc := chunk.GetCompareFunc(tp)
+	lAvgPerVal := lSide.notNullRowCount() / float64(lSide.NDV)
+	rAvgPerVal := rSide.notNullRowCount() / float64(rSide.NDV)
+	totCnt := int64(0)
+	totNdv := float64(0)
+	newHist := NewHistogram(0, 0, 0, 0, tp, 256, 0)
+	for i, j := 0, 0; i < lSide.Bounds.NumRows() && j < rSide.Bounds.NumRows(); {
+		lLow := lSide.Bounds.GetRow(i)
+		lHigh := lSide.Bounds.GetRow(i + 1)
+		rLow := rSide.Bounds.GetRow(j)
+		rHigh := rSide.Bounds.GetRow(j + 1)
+		// If [lLow, lHigh] is totally behind [rLow, rHigh], move r point.
+		if cmpFunc(lLow, 0, rHigh, 0) > 0 {
+			j += 2
+			continue
+		}
+		// If [rLow, rHigh] is totally behind [lLow, lHigh], move l point.
+		if cmpFunc(rLow, 0, lHigh, 0) > 0 {
+			i += 2
+			continue
+		}
+		var overlapLow, overLapHigh chunk.Row
+		if cmpFunc(lLow, 0, rLow, 0) > 0 {
+			overlapLow = lLow
+		} else {
+			overlapLow = rLow
+		}
+		if cmpFunc(lHigh, 0, rHigh, 0) > 0 {
+			overLapHigh = rHigh
+		} else {
+			overLapHigh = lHigh
+		}
+		// Calculate overlap ratio.
+		leftOverLap := lSide.calcRangeFraction(i/2, overlapLow, overLapHigh)
+		rightOverLap := rSide.calcRangeFraction(j/2, overlapLow, overLapHigh)
+		lCount := float64(lSide.bucketCount(i/2)) * leftOverLap
+		rCount := float64(rSide.bucketCount(j/2)) * rightOverLap
+		lNdv := lCount / lAvgPerVal
+		rNdv := rCount / rAvgPerVal
+		// bucketCount is lCount/lNdv * rCount/rNdv * finalNdv, where finalNdv is min(lNdv, rNdv).
+		bucketCount := lCount * rCount / math.Max(lNdv, rNdv)
```

Comment:
how about setting `bucketCount` to `lAvgPerVal * rAvgPerVal * math.Min(lNdv, rNdv)`?